tests/check_viterbi.c

```diff
@@ -9,83 +9,91 @@
 
 int compare_files(FILE *f1, FILE *f2)
 {
-  char ch1, ch2;
-  int flag = 0;
-
-  ch1 = fgetc(f1);
-  ch2 = fgetc(f2);
-
-  while (ch1 != EOF && ch2 != EOF) {
-    if (ch1 != ch2) {
-      flag = 1;
-      return flag;
-    }
-  ch1 = fgetc(f1);
-  ch2 = fgetc(f2);
-  }
-  if (ch1 == EOF && ch2 != ch1) {
-    flag = 1;
-    return flag;
-  }
+  int c1, c2;
+  int count = 0;
+
+  while (1) {
+    c1 = fgetc(f1);
+    c2 = fgetc(f2);
+
+    if (c1 != c2)
+      return count;
 
-  if (ch2 == EOF && ch1 != ch2) {
-    flag = 1;
-    return flag;
+    if (c1 == EOF)
+      return count;
+
+    count++;
   }
 
-  return flag;
+  return count;
 }
 
 START_TEST(test_viterbi27)
 {
-  int i;
-  struct v27 vp;
-  FILE *waas_data = NULL;
-  FILE *tmp = NULL;
-  FILE *check = NULL;
-  decision_t decisions[250 * 6];
-
-  int output = open("tmp.bin", O_WRONLY | O_CREAT, 0644);
-  waas_data = fopen("waas_data.bin", "r");
-  check = fopen("waas_check.bin", "r");
-
-  fseek(waas_data, 0, SEEK_END);
-  long fsize = ftell(waas_data);
-  fseek(waas_data, 0, SEEK_SET);
-
-  int nsymbols = fsize;
-  int nbits = nsymbols/2 - 12;
-
-  unsigned char data[nbits];
-  memset(data, 0x42, nbits);
-
-  unsigned char *buffer = (unsigned char *)malloc((fsize + 1) * sizeof(unsigned char));
-
-  fread(buffer, fsize, 1, waas_data);
-  fclose(waas_data);
-
-  for(i = 0; i < fsize; i++) {
-    char c = buffer[i];
-    int digit = c - '0';
-    if (digit == 1)
-      buffer[i] = 0xff;
-    else
-      buffer[i] = 0x00;
+  #define HISTORY_LENGTH_BITS 64
+  #define DECODE_LENGTH_BITS 32
+
+  FILE *waas_symbols = fopen("v27_sym_waas.bin", "r");
+  FILE *waas_bits = fopen("v27_bits_waas.bin", "r");
+  const char *tmp_file = "tmp.bin";
+
+  /* Read in viterbi symbols from file */
+  fseek(waas_symbols, 0, SEEK_END);
+  long fsize = ftell(waas_symbols);
+  fseek(waas_symbols, 0, SEEK_SET);
+
+  unsigned char *symbol_buffer = (unsigned char *)malloc((fsize + 1) *
+                                                         sizeof(unsigned char));
+  fread(symbol_buffer, fsize, 1, waas_symbols);
+  fclose(waas_symbols);
+
+  /* One symbol per byte in input file */
+  int nbits = fsize/2;
+
+  /* Initialize viterbi decoder */
+  v27_t v;
+  v27_decision_t decisions[HISTORY_LENGTH_BITS];
+  v27_poly_t v27_poly;
+  signed char poly_bytes[] = {V27POLYA, V27POLYB};
+  int v27_bit_length = -32; /* hack to process 32 bits before outputting data */
+
+  v27_poly_init(&v27_poly, poly_bytes);
+  v27_init(&v, decisions, HISTORY_LENGTH_BITS, &v27_poly, 0);
+
+  /* Perform decoding, writing output to file */
+  int output = open(tmp_file, O_WRONLY | O_CREAT, 0644);
```

Comment:
Why do we write the output to a file and then compare files, rather than just using memory buffers?